### PR TITLE
Feature/#197 middleware authentication

### PR DESCRIPTION
**Pull request by nachoa88**
- Refactor AuthController and UserService in order to use Passport Token.
- Modified and corrected tests related to this functions.
- Created UserPolicy in order to verify the access of an user to it's own resources. 
- Created a personalized middleware EnsureStudentOwner as an option to apply instead of the policy.
- Created tests to mock passport, middleware and policy.

### CODE REVIEW

File: tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+use Illuminate\Support\Facades\Route;
+
+
+class EnsureStudentOwnerMiddlewareTest extends TestCase
+{
+
+    use DatabaseTransactions;
+
+    public function testNonOwnersAreRedirected()
```
Comment (jordimorillo) [inline tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php:23]: testCanRedirectNonOwners

File: tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+            return 'Success';
+        })->middleware(EnsureStudentOwner::class);
+
+        // Simulate a request with a mismatched parameter
+        $response = $this->put($address);
+
+        $response->assertStatus(403);
+    }
+
+    public function testOwnersAreAllowed()
```
Comment (jordimorillo) [inline tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php:47]: testCanAllowOwners

File: tests/Feature/Policy/UserPolicyTest.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+use App\Policies\UserPolicy;
+
+use Illuminate\Auth\Access\Response;
+
+class UserPolicyTest extends TestCase
+{
+
+	use DatabaseTransactions;
+
+	public function testNonOwnersAreRestricted()
```
Comment (jordimorillo) [inline tests/Feature/Policy/UserPolicyTest.php:23]: testCanRestrictNonOwners

File: tests/Feature/Policy/UserPolicyTest.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+		$student = Student::factory()->create(['user_id' => $user->id, 'id' => '1']);
+		$student_2 = Student::factory()->create(['user_id' => $user_2->id, 'id' => '2']);
+
+		$policy = new UserPolicy();
+
+		$this->assertEquals($policy->canAccessResource($user, $user_2), Response::deny('No tens els permisos per accedir a aquest recurs.'));
+	}
+
+
+	public function testOwnersAreAllowed()
```
Comment (jordimorillo) [inline tests/Feature/Policy/UserPolicyTest.php:37]: testCanAllowOwners

File: tests/Feature/Service/Student/StudentServiceTest.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+        parent::setUp();
+        $this->studentService = new StudentService();
+        // Disable foreign key checks
+        DB::statement('SET FOREIGN_KEY_CHECKS=0;');
+    }
+
+    /**
+     * @dataProvider findUserByStudentIDSuccessProvider
+     */
+    public function testFindUserByStudentIDSuccess(string $studentID, string $userID)
```
Comment (jordimorillo) [inline tests/Feature/Service/Student/StudentServiceTest.php:32]: testCanSuccessFindingAUserByStudentId

File: tests/Feature/Service/Student/StudentServiceTest.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        return [
+            ['studentID1', 'userID1'],
+            ['studentID2', 'userID2'],
+        ];
+    }
+
+    /**
+     * @dataProvider findUserByStudentIDNotFoundProvider
+     */
+    public function testFindUserByStudentIDNotFound(string $studentID)
```
Comment (jordimorillo) [inline tests/Feature/Service/Student/StudentServiceTest.php:54]: testCanReturnNotFoundWhenFindingAUserByStudentId

File: tests/Feature/Service/Student/StudentServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        return [
+            ['nonExistingStudentID1'],
+            ['nonExistingStudentID2'],
+        ];
+    }
+
+    /**
+     * @dataProvider findUserByStudentIDUserNotFoundProvider
+     */
+    public function testFindUserByStudentIDUserNotFound(string $studentID, string $userID)
```
Comment (jordimorillo) [inline tests/Feature/Service/Student/StudentServiceTest.php:71]: testCan...

File: app/Service/Student/StudentService.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+use App\Models\Student;
+use Exception;
+
+use App\Exceptions\UserNotFoundException;
+use App\Exceptions\StudentNotFoundException;
+
+class StudentService
+{
+
+	public function __construct()
```
Comment (jordimorillo) [inline app/Service/Student/StudentService.php:17]: Si no se usa puedes eliminarlo

File: app/Service/User/UserService.php [hunk 9]
```diff
@@ -1,135 +1,88 @@
...
+		if (!$jwt) {
+			throw new CouldNotCreateJWTokenPassportException($user->id);
+		}
 		return $jwt;
 	}
 
-	
-
-	public function storeUserIDAndTokenRedis( string | int $userID, string $token): bool
+	public function storeUserIDAndTokenRedis(string | int $userID, string $token): bool
```
Comment (jordimorillo) [inline app/Service/User/UserService.php:57]: string $userID

File: app/Service/User/UserService.php [hunk 10]
```diff
@@ -1,135 +1,88 @@
...
-		}catch (Exception $e){
-			return False;
+			return true;
+		} catch (Exception $e) {
+			throw new UserNotStoredInRedisException($userID);
 		}
 	}
 
-	public function getJWTokenByUserID(string | int $userID): string | int | bool
+	public function getJWTokenByUserID(string | int $userID): string
```
Comment (jordimorillo) [inline app/Service/User/UserService.php:74]: same

**Review by jordimorillo — APPROVED** [review]
Congratulations! Fix the little things and move forward.

File: tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php (hunk 1 above)
Comment (nachoa88) [reply tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php]: Done!

File: tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php (hunk 2 above)
Comment (nachoa88) [reply tests/Feature/Middleware/EnsureStudentOwnerMiddlewareTest.php]: Done!

File: tests/Feature/Policy/UserPolicyTest.php (hunk 3 above)
Comment (nachoa88) [reply tests/Feature/Policy/UserPolicyTest.php]: Done!

File: tests/Feature/Policy/UserPolicyTest.php (hunk 4 above)
Comment (nachoa88) [reply tests/Feature/Policy/UserPolicyTest.php]: Done

File: tests/Feature/Service/Student/StudentServiceTest.php (hunk 5 above)
Comment (nachoa88) [reply tests/Feature/Service/Student/StudentServiceTest.php]: Done

File: tests/Feature/Service/Student/StudentServiceTest.php (hunk 7 above)
Comment (nachoa88) [reply tests/Feature/Service/Student/StudentServiceTest.php]: Done!

File: app/Service/User/UserService.php (hunk 9 above)
Comment (nachoa88) [reply app/Service/User/UserService.php]: Done and removed the if statement to check if ID is_numeric, no longer needed.

File: app/Service/User/UserService.php (hunk 10 above)
Comment (nachoa88) [reply app/Service/User/UserService.php]: Same as above.

File: app/Service/Student/StudentService.php (hunk 8 above)
Comment (nachoa88) [reply app/Service/Student/StudentService.php]: Deleted.

File: tests/Feature/Service/Student/StudentServiceTest.php (hunk 6 above)
Comment (nachoa88) [reply tests/Feature/Service/Student/StudentServiceTest.php]: Done!

**Review by ivanlegranbizarro — APPROVED** [review]

**Pull request merged**